File: everybody_codes/event2025/quest_05.py
"""Everyone Codes Day N."""

import collections
from typing import cast
from lib import helpers
from lib import parsers

Fishbone = collections.namedtuple("Fishbone", ["index", "segments"])


def parse(numbers: list[int]) -> Fishbone:
    """Parse a Fishbone from a line of input."""
    segments: list[list[int | None]] = []
    index, *values = numbers
    for number in values:
        for segment in segments:
            if number < cast(int, segment[1]) and segment[0] is None:
                segment[0] = number
                break
            if number > cast(int, segment[1]) and segment[2] is None:
                segment[2] = number
                break
        else:
            segments.append([None, number, None])

    return Fishbone(index, segments)


def spine(fishbone: Fishbone) -> int:
    """Extract the spine of a fishbone."""
    return int("".join(str(s[1]) for s in fishbone.segments))


def sword_sort(sword: Fishbone) -> tuple[int, list[int], int]:
    """Return sort value for a fishbone."""
    return (
        spine(sword),
        [int("".join(str(i) for i in segment if i is not None)) for segment in sword.segments],
        sword.index,
    )


def solve(part: int, data: list[list[int]]) -> int:
    """Solve the parts."""
    swords = [parse(numbers) for numbers in data]
    swords.sort(key=sword_sort, reverse=True)
    spines = [spine(i) for i in swords]
    return [
        spines[0],
        max(spines) - min(spines),
        sum(idx * pos for idx, (pos, segments) in enumerate(swords, start=1)),
    ][part - 1]


PARSER = parsers.parse_ints
TEST_DATA = [
    """\
1:2,4,1,1,8,2,7,9,8,6
2:7,9,9,3,8,3,8,8,6,8
3:4,7,6,9,1,8,3,7,2,2
4:6,4,2,1,7,4,5,5,5,8
5:2,9,3,8,3,9,5,2,1,4
6:2,4,9,6,7,4,1,7,6,8
7:2,3,7,6,2,2,4,1,4,2
8:5,1,5,6,8,3,1,8,3,9
9:5,7,7,3,7,2,3,8,6,7
10:4,1,9,3,8,5,4,3,5,5""",
    """\
1:7,1,9,1,6,9,8,3,7,2
2:6,1,9,2,9,8,8,4,3,1
3:7,1,9,1,6,9,8,3,8,3
4:6,1,9,2,8,8,8,4,3,1
5:7,1,9,1,6,9,8,3,7,3
6:6,1,9,2,8,8,8,4,3,5
7:3,7,2,2,7,4,4,6,3,1
8:3,7,2,2,7,4,4,6,3,7
9:3,7,2,2,7,4,1,6,3,7""",
    """\
1:7,1,9,1,6,9,8,3,7,2
2:7,1,9,1,6,9,8,3,7,2""",
]
TESTS = [
    (1, "58:5,3,7,8,9,10,4,5,7,8,8", 581078),
    (2, TEST_DATA[0], 77053),
    (3, TEST_DATA[1], 260),
    (3, TEST_DATA[2], 4),
]

if __name__ == "__main__":
    helpers.run_solution(globals())

File: everybody_codes/go/e2025/quest_05.go
package e2025

import (
	"cmp"
	"slices"
	"strconv"
	"strings"

	"isaacgood.com/everybodycodes/helpers"
)

// Quest05 for event 25.
type Quest05 struct{}

// Fishbone is a sword structure.
type Fishbone struct {
	index    int
	segments []*struct{ left, middle, right int }
	spine    int
	rows     []int
}

func (f Fishbone) getRows() []int {
	var rows []int
	for _, segment := range f.segments {
		var parts []string
		for _, n := range []int{segment.left, segment.middle, segment.right} {
			if n != 0 {
				parts = append(parts, helpers.Itoa(n))
			}
		}
		rows = append(rows, helpers.Atoi(strings.Join(parts, "")))
	}
	return rows
}

func (f Fishbone) getSpine() int {
	var parts []string
	for _, segment := range f.segments {
		parts = append(parts, helpers.Itoa(segment.middle))
	}
	return helpers.Atoi(strings.Join(parts, ""))
}

func newFishbone(line string) Fishbone {
	parts := strings.Split(strings.Trim(line, "\n"), ":")
	numbers := strings.Split(parts[1], ",")
	var segments []*struct{ left, middle, right int }
	for _, numRaw := range numbers {
		num := helpers.Atoi(numRaw)
		updated := false
		for _, segment := range segments {
			if num < segment.middle && segment.left == 0 {
				segment.left = num
				updated = true
				break
			} else if num > segment.middle && segment.right == 0 {
				segment.right = num
				updated = true
				break
			}
		}
		if !updated {
			segments = append(segments, &struct{ left, middle, right int }{middle: num})
		}
	}
	f := Fishbone{helpers.Atoi(parts[0]), segments, 0, nil}
	f.spine = f.getSpine()
	f.rows = f.getRows()
	return f
}

// Solve solves one part.
func (q Quest05) Solve(data string, part int) string {
	var swords []Fishbone
	var spines []int
	for line := range strings.Lines(data) {
		sword := newFishbone(line)
		swords = append(swords, sword)
		spines = append(spines, sword.spine)
	}
	switch part {
	case 1:
		return helpers.Itoa(swords[0].spine)
	case 2:
		return helpers.Itoa(slices.Max(spines) - slices.Min(spines))
	}
	slices.SortFunc(swords, func(a, b Fishbone) int {
		if i := cmp.Compare(a.spine, b.spine); i != 0 {
			return -i
		}
		for i := 0; i < slices.Min([]int{len(a.rows), len(b.rows)}); i++ {
			if i := cmp.Compare(a.rows[i], b.rows[i]); i != 0 {
				return -i
			}
		}
		return -cmp.Compare(a.index, b.index)
	})
	total := 0
	for idx, sword := range swords {
		total += (idx + 1) * sword.index
	}
	return strconv.Itoa(total)
}

func init() {
	Puzzles[5] = Quest05{}
}
